### PR TITLE
Contribute module payment_paybox (Paybox payment acquirer) to OCA

### PR DESCRIPTION
**Pull request by jbq**

### CODE REVIEW

**Comment by jbq** [comment]
@oca-clabot you're wrong I already signed my CLA

**Comment by jbq** [comment]
Initialization Error:

```
  File "/home/travis/build/OCA/account-payment/payment_paybox/models/util.py", line 5, in <module>
    from Crypto.Signature import PKCS1_v1_5
ImportError: No module named Signature
```

Seems like the OCA bot does not have python-crypto installed

**Comment by eLBati** [comment]
Hi @jbq 
you have to add it to `.travis.yml`, like
https://github.com/OCA/report-print-send/blob/8.0/.travis.yml#L28

File: payment_paybox/__init__.py [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# flake8: noqa
+
+import controllers
+import models
```
Comment (JayVora-SerpentCS) [inline payment_paybox/__init__.py]: Use v8 std. 
from . import models

File: payment_paybox/models/account_invoice.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
+# coding: utf-8
+from openerp.osv import osv, fields
+from datetime import datetime
```
Comment (JayVora-SerpentCS) [inline payment_paybox/models/account_invoice.py]: it would be appreciable if you manage to convert to api v8!
models.Model

**Comment by dreispt** [comment]
@eLBati does MQT still support a requirements.txt file?

**Comment by petrus-v** [comment]
Hi @jbq,

Thanks to take time to purpose this module to OCA.

I wonder if you could push commit histories including @fjouatte commits? I don't know what the community think about?

To do so you can do something like that (not tested)

``` bash
mkdir tmp
cd tmp
git clone https://github.com/jbq/payment_paybox
cd payment_paybox/
mkdir payment_paybox
git mv ./* payment_paybox
git commit "prepare PR"
cd ..
git clone https://github.com/OCA/account-payment
cd account-payment
git remote add paybox ../payment_paybox
git fetch paybox
git merge paybox
```

regards,

**Comment by jbq** [comment]
@petrus-v good idea, I used `git subtree` to merge both repositories, thus preserving history of the initial project.

**Comment by petrus-v** [comment]
@jbq,

Great, thanks! I didn't know we could use `subtree` in this case!

**Comment by pedrobaeza** [comment]
Thanks for the huge work of providing this module also with tests. As all commits are from 2 persons, I think you can squash all of them in 2 groups: one with one author and the other with the other person. You will have also to rebase, because there's a conflict in your branch and the main one (I suppose because you're modifying some shared files).

**Comment by pedrobaeza** [comment]
Any news about this PR? Apart from squashing commits, can you rebase?

**Comment by jbq** [comment]
Hi @pedrobaeza it took me some time to get back to this!  Everything should be fine now.  Tell me if anything is missing I'd be glad to finish this integration to the OCA.

**Comment by pedrobaeza** [comment]
You should rebase your branch as there are conflicts unresolved.

**Comment by jbq** [comment]
Yes, done

File: payment_paybox/README.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
+## Paybox payment acquirer module for Odoo ##
```
Comment (pedrobaeza) [inline payment_paybox/README.md]: Can you please make the README in RST format (the only one supported by Odoo), and using this template: https://raw.githubusercontent.com/OCA/maintainer-tools/master/template/module/README.rst?

File: payment_paybox/__openerp__.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    v3 or later along with this program.
+#    If not, see <http://www.gnu.org/licenses/>.
+#
+# ==============================================================================
+
+{
+    'name': 'Paybox payment acquirer',
+    'version': '1.0',
```
Comment (pedrobaeza) [inline payment_paybox/__openerp__.py]: Use recommended version numbering scheme: 8.0.1.0.0 - https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#version-numbers

File: payment_paybox/__openerp__.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    'sequence': 150,
+    'category': 'Custom',
+    'author': 'Odoo Community Association (OCA)',
```
Comment (pedrobaeza) [inline payment_paybox/__openerp__.py]: You should put here for credits. You can put multiple authors split by commas.

File: payment_paybox/__openerp__.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    'summary': "Payment acquirer: Paybox implementation",
+    'description': """Paybox payment acquirer, only useful for French organizations""",
+    'depends': [
+        'payment',
+        'l10n_fr'
```
Comment (pedrobaeza) [inline payment_paybox/__openerp__.py]: Why are you depending on l10n_fr? This automatically restricts the use of Paybox to any other country than France.

File: payment_paybox/__openerp__.py [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+# ==============================================================================
+
+{
+    'name': 'Paybox payment acquirer',
+    'version': '1.0',
+    'sequence': 150,
+    'category': 'Custom',
+    'author': 'Odoo Community Association (OCA)',
+    'summary': "Payment acquirer: Paybox implementation",
+    'description': """Paybox payment acquirer, only useful for French organizations""",
```
Comment (pedrobaeza) [inline payment_paybox/__openerp__.py]: This key should be removed, as you are putting README.rst that serves as description

File: payment_paybox/__openerp__.py [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        'payment',
+        'l10n_fr'
+    ],
+    'data': [
+        'templates/payment_acquirer.xml',
+        'data/payment_acquirer.xml',
+        'data/email_template.xml',
+        'views/payment_acquirer.xml',
+    ],
+    'demo_xml': [
```
Comment (pedrobaeza) [inline payment_paybox/__openerp__.py]: This one and init_xml are deprecated keys. Please remove it (and you can also do the same for empty ones).

File: payment_paybox/__openerp__.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    ],
+    'init_xml': [
+    ],
+    'css': [],
+    'test': [],
+    'installable': True,
+    'application': True,
```
Comment (pedrobaeza) [inline payment_paybox/__openerp__.py]: This is not an application per se

File: payment_paybox/__openerp__.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+    'auto_install': False,
+    'license': 'AGPL-3',
+    'post_load': None,
```
Comment (pedrobaeza) [inline payment_paybox/__openerp__.py]: What is this key? You have post_init_hook, but not post_load. Anyway, it's not used, so please remove it.

File: payment_paybox/controllers/__init__.py [hunk 11]
```diff
@@ -0,0 +1,3 @@
+# flake8: noqa
```
Comment (pedrobaeza) [inline payment_paybox/controllers/__init__.py]: You don't need to exclude it from flake8. It's already handled on current Travis configuration

File: payment_paybox/__openerp__.py [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+#    v3 or later along with this program.
+#    If not, see <http://www.gnu.org/licenses/>.
+#
+# ==============================================================================
+
+{
+    'name': 'Paybox payment acquirer',
+    'version': '1.0',
+    'sequence': 150,
+    'category': 'Custom',
```
Comment (pedrobaeza) [inline payment_paybox/__openerp__.py]: Use the same category as main payment module

File: payment_paybox/data/email_template.xml [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+        <record id="paybox_error_reconcile_email_template" model="email.template">
+            <field name="name">Warning Reconcile Paybox</field>
+            <field name="email_from">${user.email}</field>
+            <field name="subject">[Paybox] Avertissement lettrage</field>
+            <field name="email_to">${object.get_paybox_admin()}</field>
+            <field name="lang">${user.lang}</field>
+            <field name="model_id" ref="payment_paybox.model_paybox_warning"/>
+            <field name="auto_delete" eval="True"/>
+            <field name="body_html">
+                Une erreur s'est produite lors de la tentative de validation d'un paiement Paybox.
```
Comment (pedrobaeza) [inline payment_paybox/data/email_template.xml]: Can you please put this on English and then translated on fr.po?

File: payment_paybox/models/account_invoice.py [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+        today = datetime.today()
+        bank_journal_id = journal.search(cr, uid, [('code', '=', 'BNK2')])
+        period_id = invoice.period_id.id if invoice else period.find(cr, uid, today)[0]
+        values = {'journal_id': bank_journal_id[0], 'period_id': period_id,
+                  'date': today, 'name': '/'}
+        move_id = move.create(cr, uid, values)
+        return move_id
+
+    def create_move_lines(self, cr, uid, invoice, move_id, montant):
+        """ create move lines for the given amount and linked to given move_id """
```
Comment (pedrobaeza) [inline payment_paybox/models/account_invoice.py]: Why are you creating these move lines? payment modules aren't supposed to create it, but only handle the communication itself with the payment acquirer. Maybe you should create a bridge module (that depends on l10n_fr) that facilitates the accounting with this module in France, but in a split way

**Comment by jbq** [comment]
@JayVora-SerpentCS your contribution will be welcome to port to the new API.  I can't volunteer more work on this project.

**Comment by jbq** [comment]
@pedrobaeza I updated the PR according to your feedback

File: payment_paybox/README.rst [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+This modules installs with default values so that you can reach the Paybox
+payment screen.  For proper tests, also install the ``website_sale`` module.  Go
+to the Payment Acquirers configuration screen to set the return URL of your
+Odoo installation to properly process the payment.
+
+Bug Tracker
+===========
+
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/
+{project_repo}/issues>`_.
```
Comment (pedrobaeza) [inline payment_paybox/README.rst]: You should replace `{project_repo}` and the other place_holders with the proper values

File: payment_paybox/README.rst [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback `here <https://github.com/OCA/
+{project_repo}/issues/new?body=module:%20
+{module_name}%0Aversion:%20
+{version}%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
```
Comment (pedrobaeza) [inline payment_paybox/README.rst]: version is Odoo version (8.0)

**Comment by pedrobaeza** [comment]
Thank you very much for all the efforts. Little remarks, but overall the code review is :+1: 

I haven't tested personally, but the test gives me confidence on that.

**Comment by jbq** [comment]
Hi, can you please merge this PR before the code gets obsolete?

**Comment by pedrobaeza** [comment]
There's still errors on Travis. You should check them. And please write a comment when you make changes to warn all the subscribers of this PR.

**Comment by jbq** [comment]
About this import error:

```
    from Crypto.Signature import PKCS1_v1_5
ImportError: No module named Signature
```

I already added `python-crypto` requirement in the Travis configuration, but it looks like it is ignored.  Can you please help?

**Comment by pedrobaeza** [comment]
It seems that python-crypto doesn't exist as pre-configured package. Try to put on install section `pip install crypto`

**Comment by jbq** [comment]
@pedrobaeza validating all the tests looks like a daunting task, I won't be able to invest more time on it.  Is there a way to squeeze the tests for this particular module?  I'd like it to be included as is.  Thanks for your help!

**Comment by pedrobaeza** [comment]
Sorry, but you have to fix at least PEP8 problems (see LINT_CHECK section), and solve the library problem

**Comment by pedrobaeza** [comment]
Closing as no further answer and Travis is still red.

**Pull request closed without merging**